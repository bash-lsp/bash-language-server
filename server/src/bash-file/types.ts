import * as LSP from 'vscode-languageserver'

export type Kinds = { [type: string]: LSP.SymbolKind }

export type Declarations = { [name: string]: LSP.SymbolInformation[] }

export interface BaseBashFile {
  problems: LSP.Diagnostic[]
  declarations: Declarations

  /**
   * Find all occurrences of name in the given file.
   * It's currently not scope-aware.
   */
  findOccurrences(uri: string, query: string): LSP.Location[]

  /**
   * Find the node at the given point.
   *
   * TODO: should be private as it exposes the syntax node
   */
  nodeAtPoint(line: number, column: number): any | null

  /**
   * Find the full word at the given point.
   */
  wordAtPoint(line: number, column: number): string | null

  /**
   * Find the name of the command at the given point.
   */
  commandNameAtPoint(line: number, column: number): string | null
}
