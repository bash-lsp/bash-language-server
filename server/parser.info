"https://api.github.com/repos/tree-sitter/tree-sitter-bash/git/commits/4488aa41406547e478636a4fcfd24f5bbc3f2f74"
tree-sitter-cli "0.20.7"
