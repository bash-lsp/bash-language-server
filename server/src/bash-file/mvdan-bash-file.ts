import type { File as FileNode, Node as SyntaxNode } from 'mvdan-sh'
import * as LSP from 'vscode-languageserver'
import { Range } from 'vscode-languageserver/lib/main'

const ShellScript = require('mvdan-sh') // NOTE: import is not working
const _parser = ShellScript.syntax.NewParser()

import { BaseBashFile, Declarations, Kinds } from './types'

const syntaxNodeTypeToLSPKind: Kinds = {
  FuncDecl: LSP.SymbolKind.Function,
  Assign: LSP.SymbolKind.Variable,
}

function createRange(node: SyntaxNode): Range {
  return Range.create(
    node.Pos().Line() - 1,
    node.Pos().Col() - 1,
    node.End().Line() - 1,
    node.End().Col() - 1,
  )
}

interface ValueNode extends SyntaxNode {
  Value: string
}

interface NamedNode extends SyntaxNode {
  Name: ValueNode
}

function isDefinition(node: SyntaxNode): node is NamedNode {
  switch (ShellScript.syntax.NodeType(node)) {
    // TODO: other cases?
    case 'Assign':
      return (node as any).Name !== null
    case 'FuncDecl':
      return true
    default:
      return false
  }
}

/*
function traverseParsed(node: Sh.ParsedSh, act: (node: Sh.ParsedSh) => void) {
  act(node)
  getChildren(node).forEach((child) => traverseParsed(child, act))
}

function withParents<T extends Sh.ParsedSh>(root: T) {
  traverseParsed(root, (node) => {
    getChildren(node).forEach((child) => ((child as Sh.BaseNode).parent = node))
  })
  return root
}
*/

/**
 * The Analyzer uses the Abstract Syntax Trees (ASTs) that are provided by
 * tree-sitter to find definitions, reference, etc.
 */
export class BashFile implements BaseBashFile {
  private fileNode: FileNode
  public problems: LSP.Diagnostic[]
  private contents: string
  public declarations: Declarations

  public constructor(
    fileNode: FileNode,
    problems: LSP.Diagnostic[],
    contents: string,
    declarations: Declarations,
  ) {
    this.fileNode = fileNode
    this.problems = problems
    this.contents = contents
    this.declarations = declarations
  }

  public static parse({
    uri,
    contents,
    parser,
  }: {
    uri: string
    contents: string
    parser: any
  }): BashFile {
    const fileNode = _parser.Parse(contents)

    const problems: LSP.Diagnostic[] = []
    const declarations: Declarations = {}

    // TODO: handle errros

    //ShellScript.syntax.DebugPrint(fileNode)

    ShellScript.syntax.Walk(fileNode, (node: SyntaxNode) => {
      if (!node) {
        return true
      }

      if (isDefinition(node)) {
        const name = node.Name.Value
        const namedDeclarations = declarations[name] || []

        /*
        const parent = findParent(n, (p) => p.type === 'function_definition')
        const parentName =
          parent && parent.firstNamedChild
            ? contents.slice(
                parent.firstNamedChild.startIndex,
                parent.firstNamedChild.endIndex,
              )
            : '' // TODO: unsure what we should do here?
        */

        // TODO: parent?
        const containerName = undefined

        namedDeclarations.push(
          LSP.SymbolInformation.create(
            name,
            syntaxNodeTypeToLSPKind[ShellScript.syntax.NodeType(node)],
            createRange(node),
            uri,
            containerName,
          ),
        )
        declarations[name] = namedDeclarations
      }

      return true
    })

    return new BashFile(fileNode, problems, contents, declarations)
  }

  /**
   * Find all occurrences of name in the given file.
   * It's currently not scope-aware.
   */
  public findOccurrences(uri: string, query: string): LSP.Location[] {
    const locations: LSP.Location[] = []

    ShellScript.syntax.Walk(this.fileNode, (node: SyntaxNode) => {
      if (!node) {
        return true
      }

      let relevantNode: ValueNode | undefined

      if (isDefinition(node)) {
        relevantNode = node.Name
      } else if (ShellScript.syntax.NodeType(node) === 'ParamExp') {
        relevantNode = (node as any).Param
      }

      if (relevantNode?.Value === query) {
        const range = createRange(relevantNode)
        locations.push(LSP.Location.create(uri, range))
      }

      return true
    })

    return locations
  }

  /**
   * Find the node at the given point.
   *
   * TODO: should be private as it exposes the syntax node
   */
  public nodeAtPoint(line: number, column: number): SyntaxNode | null {
    throw new Error('not implemented')
    // return document.rootNode.descendantForPosition({ row: line, column })
  }

  /**
   * Find the full word at the given point.
   */
  public wordAtPoint(line: number, column: number): string | null {
    throw new Error('not implemented')
    /*
    const node = this.nodeAtPoint(line, column)

    if (!node || node.childCount > 0 || node.text.trim() === '') {
      return null
    }

    return node.text.trim()
    */
  }

  /**
   * Find the name of the command at the given point.
   */
  public commandNameAtPoint(line: number, column: number): string | null {
    throw new Error('not implemented')
    /*
    let node = this.nodeAtPoint(line, column)

    while (node && node.type !== 'command') {
      node = node.parent
    }

    if (!node) {
      return null
    }

    const firstChild = node.firstNamedChild

    if (!firstChild || firstChild.type !== 'command_name') {
      return null
    }

    return firstChild.text.trim()
    */
  }
}
