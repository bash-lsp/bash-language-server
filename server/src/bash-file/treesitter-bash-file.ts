import * as LSP from 'vscode-languageserver'
import { Range } from 'vscode-languageserver/lib/main'
import * as Parser from 'web-tree-sitter'
import { SyntaxNode } from 'web-tree-sitter'

import { BaseBashFile, Declarations, Kinds } from './types'

const treeSitterTypeToLSPKind: Kinds = {
  // These keys are using underscores as that's the naming convention in tree-sitter.
  environment_variable_assignment: LSP.SymbolKind.Variable,
  function_definition: LSP.SymbolKind.Function,
  variable_assignment: LSP.SymbolKind.Variable,
}

function forEach(node: SyntaxNode, cb: (n: SyntaxNode) => void) {
  cb(node)
  if (node.children.length) {
    node.children.forEach((n) => forEach(n, cb))
  }
}

function createRange(n: SyntaxNode): Range {
  return Range.create(
    n.startPosition.row,
    n.startPosition.column,
    n.endPosition.row,
    n.endPosition.column,
  )
}

function isDefinition(n: SyntaxNode): boolean {
  switch (n.type) {
    // For now. Later we'll have a command_declaration take precedence over
    // variable_assignment
    case 'variable_assignment':
    case 'function_definition':
      return true
    default:
      return false
  }
}

function isReference(n: SyntaxNode): boolean {
  switch (n.type) {
    case 'variable_name':
    case 'command_name':
      return true
    default:
      return false
  }
}

function findParent(
  start: SyntaxNode,
  predicate: (n: SyntaxNode) => boolean,
): SyntaxNode | null {
  let node = start.parent
  while (node !== null) {
    if (predicate(node)) {
      return node
    }
    node = node.parent
  }
  return null
}

/**
 * The Analyzer uses the Abstract Syntax Trees (ASTs) that are provided by
 * tree-sitter to find definitions, reference, etc.
 */
export class BashFile implements BaseBashFile {
  private ast: Parser.Tree
  public problems: LSP.Diagnostic[]
  private contents: string
  public declarations: Declarations

  public constructor(
    ast: Parser.Tree,
    problems: LSP.Diagnostic[],
    contents: string,
    declarations: Declarations,
  ) {
    this.ast = ast
    this.problems = problems
    this.contents = contents
    this.declarations = declarations
  }

  public static parse({
    uri,
    contents,
    parser,
  }: {
    uri: string
    contents: string
    parser: Parser
  }): BashFile {
    const tree = parser.parse(contents)

    const problems: LSP.Diagnostic[] = []
    const declarations: Declarations = {}

    forEach(tree.rootNode, (n: Parser.SyntaxNode) => {
      if (n.type === 'ERROR') {
        problems.push(
          LSP.Diagnostic.create(
            createRange(n),
            'Failed to parse expression',
            LSP.DiagnosticSeverity.Error,
          ),
        )
        return
      } else if (isDefinition(n)) {
        const named = n.firstNamedChild

        if (named === null) {
          return
        }

        const name = contents.slice(named.startIndex, named.endIndex)
        const namedDeclarations = declarations[name] || []

        const parent = findParent(n, (p) => p.type === 'function_definition')
        const parentName =
          parent && parent.firstNamedChild
            ? contents.slice(
                parent.firstNamedChild.startIndex,
                parent.firstNamedChild.endIndex,
              )
            : '' // TODO: unsure what we should do here?

        namedDeclarations.push(
          LSP.SymbolInformation.create(
            name,
            treeSitterTypeToLSPKind[n.type],
            createRange(n),
            uri,
            parentName,
          ),
        )
        declarations[name] = namedDeclarations
      }
    })

    function findMissingNodes(node: Parser.SyntaxNode) {
      if (node.isMissing()) {
        problems.push(
          LSP.Diagnostic.create(
            createRange(node),
            `Syntax error: expected "${node.type}" somewhere in the file`,
            LSP.DiagnosticSeverity.Warning,
          ),
        )
      } else if (node.hasError()) {
        node.children.forEach(findMissingNodes)
      }
    }

    findMissingNodes(tree.rootNode)

    return new BashFile(tree, problems, contents, declarations)
  }

  /**
   * Find all occurrences of name in the given file.
   * It's currently not scope-aware.
   */
  public findOccurrences(uri: string, query: string): LSP.Location[] {
    const { ast: tree, contents } = this

    const locations: LSP.Location[] = []

    forEach(tree.rootNode, (n) => {
      let name: null | string = null
      let range: null | LSP.Range = null

      if (isReference(n)) {
        const node = n.firstNamedChild || n
        name = contents.slice(node.startIndex, node.endIndex)
        range = createRange(node)
      } else if (isDefinition(n)) {
        const namedNode = n.firstNamedChild
        if (namedNode) {
          name = contents.slice(namedNode.startIndex, namedNode.endIndex)
          range = createRange(namedNode)
        }
      }

      if (name === query && range !== null) {
        locations.push(LSP.Location.create(uri, range))
      }
    })

    return locations
  }

  /**
   * Find the node at the given point.
   *
   * TODO: should be private as it exposes the syntax node
   */
  public nodeAtPoint(line: number, column: number): Parser.SyntaxNode | null {
    const document = this.ast

    if (!document?.rootNode) {
      // Check for lacking rootNode (due to failed parse?)
      return null
    }

    return document.rootNode.descendantForPosition({ row: line, column })
  }

  /**
   * Find the full word at the given point.
   */
  public wordAtPoint(line: number, column: number): string | null {
    const node = this.nodeAtPoint(line, column)

    if (!node || node.childCount > 0 || node.text.trim() === '') {
      return null
    }

    return node.text.trim()
  }

  /**
   * Find the name of the command at the given point.
   */
  public commandNameAtPoint(line: number, column: number): string | null {
    let node = this.nodeAtPoint(line, column)

    while (node && node.type !== 'command') {
      node = node.parent
    }

    if (!node) {
      return null
    }

    const firstChild = node.firstNamedChild

    if (!firstChild || firstChild.type !== 'command_name') {
      return null
    }

    return firstChild.text.trim()
  }
}
