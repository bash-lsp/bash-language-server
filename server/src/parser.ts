import * as Parser from 'web-tree-sitter'

export async function initializeParser(): Promise<Parser> {
  await Parser.init()
  const parser = new Parser()

  /**
   * See https://github.com/tree-sitter/tree-sitter/tree/master/lib/binding_web#generate-wasm-language-files
   *
   * To compile:
   *  yarn add --dev tree-sitter-cli
   *  npx tree-sitter build-wasm node_modules/tree-sitter-bash
   *
   * The current files was compiled with:
   * "tree-sitter-bash": "^0.16.0",
   * "tree-sitter-cli": "^0.15.9"
   */
  const lang = await Parser.Language.load(`${__dirname}/../tree-sitter-bash.wasm`)

  parser.setLanguage(lang)
  return parser
}
