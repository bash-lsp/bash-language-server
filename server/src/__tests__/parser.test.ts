import FIXTURES from '../../../testing/fixtures'
//import { BashFile } from '../bash-file/treesitter-bash-file'
import { BashFile } from '../bash-file/mvdan-bash-file'
import { initializeParser } from '../parser'

let parser: any

const CURRENT_URI = 'dummy-uri.sh'

beforeAll(async () => {
  parser = await initializeParser()
})

describe('bash-file parsing', () => {
  it('returns an empty list of errors for a file with no parsing errors', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.INSTALL.getText(),
      parser,
    })
    expect(bashFile.problems).toEqual([])
  })

  it('returns a list of errors for a file with a missing node', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.MISSING_NODE.getText(),
      parser,
    })
    expect(bashFile.problems).not.toEqual([])
  })

  it('creates declarations', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.ISSUE101.getText(),
      parser,
    })
    expect(bashFile.declarations).toMatchSnapshot()
  })
})

describe('findOccurrences', () => {
  it('returns empty list if parameter is not found', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.INSTALL.getText(),
      parser,
    })
    const result = bashFile.findOccurrences(CURRENT_URI, 'foobar')
    expect(result).toEqual([])
  })

  it('returns a list of locations if parameter is found', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.INSTALL.getText(),
      parser,
    })
    const result = bashFile.findOccurrences(CURRENT_URI, 'node_version')
    expect(result).not.toEqual([])
    expect(result).toMatchSnapshot()
  })
})

describe('wordAtPoint', () => {
  it('returns current word at a given point', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.INSTALL.getText(),
      parser,
    })
    expect(bashFile.wordAtPoint(25, 0)).toEqual(null)
    expect(bashFile.wordAtPoint(25, 1)).toEqual(null)
    expect(bashFile.wordAtPoint(25, 2)).toEqual(null)
    expect(bashFile.wordAtPoint(25, 3)).toEqual(null)
    expect(bashFile.wordAtPoint(25, 4)).toEqual('rm')
    expect(bashFile.wordAtPoint(25, 5)).toEqual('rm')
    expect(bashFile.wordAtPoint(25, 6)).toEqual(null)
    expect(bashFile.wordAtPoint(25, 7)).toEqual('npm-install-')

    expect(bashFile.wordAtPoint(24, 2)).toEqual('else')
    expect(bashFile.wordAtPoint(24, 3)).toEqual('else')
    expect(bashFile.wordAtPoint(24, 5)).toEqual('else')
    expect(bashFile.wordAtPoint(24, 7)).toEqual(null)

    expect(bashFile.wordAtPoint(30, 1)).toEqual(null)

    expect(bashFile.wordAtPoint(30, 2)).toEqual('ret')
    expect(bashFile.wordAtPoint(30, 3)).toEqual('ret')
    expect(bashFile.wordAtPoint(30, 4)).toEqual('ret')
    expect(bashFile.wordAtPoint(30, 5)).toEqual('=')

    expect(bashFile.wordAtPoint(38, 5)).toEqual('configures')
  })
})

describe('commandNameAtPoint', () => {
  it('returns current command name at a given point', () => {
    const bashFile = BashFile.parse({
      uri: CURRENT_URI,
      contents: FIXTURES.INSTALL.getText(),
      parser,
    })
    expect(bashFile.commandNameAtPoint(15, 0)).toEqual(null)

    expect(bashFile.commandNameAtPoint(20, 2)).toEqual('curl')
    expect(bashFile.commandNameAtPoint(20, 15)).toEqual('curl')
    expect(bashFile.commandNameAtPoint(20, 19)).toEqual('curl')

    expect(bashFile.commandNameAtPoint(26, 4)).toEqual('echo')
    expect(bashFile.commandNameAtPoint(26, 9)).toEqual('echo')

    expect(bashFile.commandNameAtPoint(38, 13)).toEqual('env')
    expect(bashFile.commandNameAtPoint(38, 24)).toEqual('grep')
    expect(bashFile.commandNameAtPoint(38, 44)).toEqual('sed')
  })
})
