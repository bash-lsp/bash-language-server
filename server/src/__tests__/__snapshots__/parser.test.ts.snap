// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bash-file parsing creates declarations 1`] = `
Object {
  "COMMENTS": Array [
    Object {
      "containerName": "add_a_user",
      "kind": 13,
      "location": Object {
        "range": Object {
          "end": Object {
            "character": 13,
            "line": 9,
          },
          "start": Object {
            "character": 2,
            "line": 9,
          },
        },
        "uri": "dummy-uri.sh",
      },
      "name": "COMMENTS",
    },
  ],
  "PASSWORD": Array [
    Object {
      "containerName": "add_a_user",
      "kind": 13,
      "location": Object {
        "range": Object {
          "end": Object {
            "character": 13,
            "line": 6,
          },
          "start": Object {
            "character": 2,
            "line": 6,
          },
        },
        "uri": "dummy-uri.sh",
      },
      "name": "PASSWORD",
    },
  ],
  "USER": Array [
    Object {
      "containerName": "add_a_user",
      "kind": 13,
      "location": Object {
        "range": Object {
          "end": Object {
            "character": 9,
            "line": 5,
          },
          "start": Object {
            "character": 2,
            "line": 5,
          },
        },
        "uri": "dummy-uri.sh",
      },
      "name": "USER",
    },
  ],
  "add_a_user": Array [
    Object {
      "kind": 12,
      "location": Object {
        "range": Object {
          "end": Object {
            "character": 1,
            "line": 14,
          },
          "start": Object {
            "character": 0,
            "line": 3,
          },
        },
        "uri": "dummy-uri.sh",
      },
      "name": "add_a_user",
    },
  ],
}
`;

exports[`findOccurrences returns a list of locations if parameter is found 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 12,
        "line": 148,
      },
      "start": Object {
        "character": 0,
        "line": 148,
      },
    },
    "uri": "dummy-uri.sh",
  },
  Object {
    "range": Object {
      "end": Object {
        "character": 45,
        "line": 152,
      },
      "start": Object {
        "character": 33,
        "line": 152,
      },
    },
    "uri": "dummy-uri.sh",
  },
  Object {
    "range": Object {
      "end": Object {
        "character": 20,
        "line": 162,
      },
      "start": Object {
        "character": 8,
        "line": 162,
      },
    },
    "uri": "dummy-uri.sh",
  },
  Object {
    "range": Object {
      "end": Object {
        "character": 31,
        "line": 165,
      },
      "start": Object {
        "character": 19,
        "line": 165,
      },
    },
    "uri": "dummy-uri.sh",
  },
]
`;
